### PR TITLE
Improve export nncf models

### PR DESCRIPTION
**Pull request by LeonidBeynenson**
* Refactor tools/export.py and BaseDetector.export, implement BaseDetector.forward_export_context
* Move get_fake_input to a separate file
* Make more tricky wrap_nncf_model
  (now it may use forward_export instead of forward_dummy method of mmdetection model)
* Make tools/export.py export to onnx with nncf-wrapped models
* Add more accurate import of nncf classes in mmdet/core/nncf/utils.py

### CODE REVIEW

**Comment by LeonidBeynenson** [comment]
2 @aleksei-kashapov : Alexei, please, take a look

File: mmdet/core/nncf/utils.py [hunk 1]
```diff
@@ -26,18 +27,41 @@ def check_nncf_is_enabled():
...
+        class_InitializingDataLoader = InitializingDataLoader
     except:
         raise RuntimeError("Incompatible version of NNCF")
+else:
+    class DummyCompressionAlgorithmController:
+        pass
+    class DummyInitializingDataLoader:
+        pass
+
+    class_CompressionAlgorithmController = DummyCompressionAlgorithmController
```
Comment (aleksei-kashapov) [inline mmdet/core/nncf/utils.py:48]: Looks pretty overhead ...

File: mmdet/core/nncf/utils.py [hunk 2]
```diff
@@ -26,18 +27,41 @@ def check_nncf_is_enabled():
...
+    class_InitializingDataLoader = DummyInitializingDataLoader
 
+def check_NNCFNetwork_is_compatible():
+    check_nncf_is_enabled()
+    from nncf.nncf_network import NNCFNetwork
+    assert not hasattr(NNCFNetwork, "export"), "NNCFNetwork should not contain 'export', since it is defined in BaseDetector"
```
Comment (aleksei-kashapov) [inline mmdet/core/nncf/utils.py:54]: I am not sure that this assert is good enough. What about just checking the version of NNCF?

File: mmdet/core/nncf/utils.py [hunk 3]
```diff
@@ -56,13 +80,50 @@ def wrap_nncf_model(model, cfg, data_loader_for_init=None):
...
+        # NB: the full cfg is required here!
+        fake_data = get_fake_input_func(cfg, orig_img_shape=orig_img_shape, device=device)
+        return fake_data
+    def _get_fake_data_for_forward_export():
+        # make a closure to use config and get_fake_input_func from the external scope
+        cfg_copy = deepcopy(cfg)
+        nncf_config = NNCFConfig(cfg_copy.nncf_config)
+        return __get_fake_data_for_forward_export(cfg_copy, nncf_config, get_fake_input_func)
+
+    def dummy_forward_without_export_part(model):
```
Comment (aleksei-kashapov) [inline mmdet/core/nncf/utils.py:106]: I would remove this function and always use input from get_fake_input()

File: mmdet/core/nncf/compression_hooks.py [hunk 4]
```diff
@@ -21,13 +22,7 @@ def before_run(self, runner):
             print_statistics(self.compression_ctrl.statistics(), runner.logger)
 
     def after_run(self, runner):
```
Comment (aleksei-kashapov) [inline mmdet/core/nncf/compression_hooks.py:24]: Just need to completely remove this method

File: mmdet/core/nncf/compression_hooks.py [hunk 5]
```diff
@@ -7,7 +7,8 @@ def __init__(self, compression_ctrl=None, cfg=None):
         from nncf import NNCFConfig
 
         self.compression_ctrl = compression_ctrl
-        self.nncf_config = NNCFConfig(cfg.nncf_config)
+        # TODO: is it required now?
```
Comment (aleksei-kashapov) [inline mmdet/core/nncf/compression_hooks.py:10]: No, we don't use it anymore

File: mmdet/models/detectors/base.py [hunk 6]
```diff
@@ -166,18 +169,29 @@ def forward(self, img, img_metas=[], return_loss=True, **kwargs):
...
         self.forward_backup = self.forward
         self.forward = self.forward_export
-        torch.onnx.export(self, img, export_name, **kwargs)
+        yield
         self.forward = self.forward_backup
+        self.forward_backup = None
+        self.img_metas = None
+
+    def export(self, img, img_metas, export_name='', **kwargs):
+        with self.forward_export_context(img_metas):
```
Comment (aleksei-kashapov) [inline mmdet/models/detectors/base.py:193]: Why do you think using the context is better here? I mind that the previous approach was easier to understand

File: tools/test.py [hunk 7]
```diff
@@ -121,7 +134,11 @@ def main():
     # nncf model wrapper
     if 'nncf_config' in cfg:
         check_nncf_is_enabled()
-        _, model = wrap_nncf_model(model, cfg, None)
+        cfg.nncf_load_from = args.checkpoint
```
Comment (aleksei-kashapov) [inline tools/test.py:137]: Is it expected behavior that we can overwrite the config parameter by script arguments?

**Review by aleksei-kashapov — COMMENTED** [review]
I would like to have a discussion about this PR

File: mmdet/core/nncf/utils.py (hunk 3 above)
Comment (LeonidBeynenson) [reply mmdet/core/nncf/utils.py:106]: Considered, and decided to keep it as is.

File: mmdet/core/nncf/utils.py (hunk 2 above)
Comment (LeonidBeynenson) [reply mmdet/core/nncf/utils.py:54]: Considered and resolved

File: mmdet/core/nncf/compression_hooks.py (hunk 4 above)
Comment (LeonidBeynenson) [reply mmdet/core/nncf/compression_hooks.py:24]: Ok

File: mmdet/core/nncf/compression_hooks.py (hunk 5 above)
Comment (LeonidBeynenson) [reply mmdet/core/nncf/compression_hooks.py:10]: Ok

File: mmdet/models/detectors/base.py (hunk 6 above)
Comment (LeonidBeynenson) [reply mmdet/models/detectors/base.py:193]: Considered and resolved

File: tools/test.py (hunk 7 above)
Comment (LeonidBeynenson) [reply tools/test.py:137]: Yes! 🙂

**Pull request merged**